### PR TITLE
Use `insert_same` in `insert_evaluation_cache`

### LINKED ISSUE #50507: EvaluationCache overwrites its entries
When the insertion to `EvaluationCache ` is changed to use `HashMapExt::insert_same`. It triggers an the assertion which checks that the old value is the same as the new one. This is because the DepNodeIndex changed.

[Backtrace](https://gist.github.com/Zoxc/c81faf46c9b1a532f029c1a3776c6a4b#file-gistfile1-txt)
[Log output](https://gist.github.com/Zoxc/c81faf46c9b1a532f029c1a3776c6a4b#file-log-output)

cc @arielb1 @nikomatsakis

### PR DESCRIPTION
**Pull request by GabrielMajeri**
Closes #50507.

I was trying to fix this issue, but it seems the reported crash doesn't happen anymore.

### CODE REVIEW

**Comment by estebank** [comment]
The code change looks correct, but deferring to @eddyb

File: src/librustc_trait_selection/traits/select/mod.rs [hunk 1]
```diff
@@ -901,10 +901,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                     trait_ref, result,
                 );
                 // This may overwrite the cache with the same value
-                // FIXME: Due to #50507 this overwrites the different values
-                // This should be changed to use HashMapExt::insert_same
-                // when that is fixed
-                self.tcx().evaluation_cache.insert(param_env.and(trait_ref), dep_node, result);
+                self.tcx().evaluation_cache.insert_same(param_env.and(trait_ref), dep_node, result);
```
Comment (nikomatsakis) [inline src/librustc_trait_selection/traits/select/mod.rs:904]: So, #50507 is not yet closed -- but I think it is maybe specific to parallel execution? i.e., what prompted this PR?

Comment (GabrielMajeri) [reply src/librustc_trait_selection/traits/select/mod.rs:904]: I was looking for issues to work on in the list at https://github.com/rust-lang/rust/issues/48685, and found a reference to that issue. I began investigating the `panic` mentioned in the issue description, and was unable to reproduce it.

**Comment by Dylan-DPC** [comment]
@nikomatsakis can we merge this or is this blocked on #50507 being merged first?

**Comment by nikomatsakis** [comment]
Yeah, sorry, I got very distracted. I think I'd be inclined not to land this PR until #50507 is closed, just because it may cause unexpected problems if we were to enable the parallel mode. I guess the main thing would be to track down the problem -- though the bug report is pretty vague.

In any case, it's borderline, but I'm inclined to just leave the code for now since it's just one line. 

@GabrielMajeri I do appreciate the PR though! Sorry for the confusion. :(

**Pull request closed without merging**